### PR TITLE
Add datasets predicted by NRC model

### PR DESCRIPTION
**Pull request by DucPhamBP**

### CODE REVIEW

**Comment by SaraHoxha** [comment]
@DucPhamBP  Review is needed. I'll write here below my points of things to fix:
1. Your code and files are inconsistent (the paths, the file formats are not the same) and as such i couldn't run your code.
2. NRC is rule-based so we just need to run the test set.
3. You need to compute metrics about the result so NRC so that we can then compare with BiRNN.
4. These metrics should be saved into results folder.
5. Move NRC into a separate folder not inside hierarchy of Pretrained Models.
6. Is there any reason you aren't using the library NRCLex but instead implementing it yourself? Might be easier if you use the library.

Please reply to this comment when you have made the necessary changes

**Comment by DucPhamBP** [comment]
1. I checked and fixed them, it lacked something so I added and now it is okay. If you want, may I send to you to review first?
2. At the updated files, I removed the training set. It only has test set
3. I supplemented some metrics in the updated files also
6. Actually, I used NRCLex before, but the results were really poor. I checked the meaning of some texts, and some of them had the opposite emotion to the predicted emotion. Moreover, some texts could not be predicted the emotion because the words in the text were not in the model. Therefore, I tried to search for something to improve the performance. I used another version of NRC from https://saifmohammad.com/WebPages/AffectIntensity.htm. In this model, they calculate the intensity of each emotion with a score for each word, so it could be more accuracy. This is the reason I chose the NRC Emotion Intensity Lexicon instead of using NRCLex.

File: NRC/yangswei_85_nrc_intensity.ipynb [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+       "4  What are good job sites to find LEGIT remote w...   joy"
+      ]
+     },
+     "execution_count": 4,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "yangswei_85_test = pd.read_csv('test_yangswei_85.csv')\n",
```
Comment (SaraHoxha) [inline NRC/yangswei_85_nrc_intensity.ipynb]: path here is not correct

File: NRC/yangswei_85_nrc_intensity.ipynb [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "a3f458c0-ef02-4e30-a2d0-e7562b4fc5a3",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "yangswei_85_test.to_csv('yangswei_85_nrc_inten_test.csv')"
```
Comment (SaraHoxha) [inline NRC/yangswei_85_nrc_intensity.ipynb]: here the format is 'csv' but the uploaded files are in xls. why is this discrepancy? it's better we use csv.

File: NRC/yangswei_85_nrc_intensity.ipynb [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+   ],
+   "source": [
+    "from sklearn.metrics import accuracy_score, precision_recall_fscore_support\n",
+    "y_true = yangswei_85_test['label']\n",
+    "y_pred = yangswei_85_test['label_nrc']\n",
+    "\n",
+    "accuracy = accuracy_score(y_true, y_pred)\n",
+    "precision, recall, f1, _ = precision_recall_fscore_support(y_true, y_pred, average='weighted')\n",
+    "\n",
+    "# show the results\n",
```
Comment (SaraHoxha) [inline NRC/yangswei_85_nrc_intensity.ipynb]: add code that saves the results to a file, from the notebook there is no automation regarding the saving of files.

File: NRC/t5_nrc_intensity.ipynb [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+       "4  There’s no magic formula to get remote work so...   fear"
+      ]
+     },
+     "execution_count": 9,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "t5_test = pd.read_csv('test_t5.csv')\n",
```
Comment (SaraHoxha) [inline NRC/t5_nrc_intensity.ipynb]: file path not correct

File: NRC/t5_nrc_intensity.ipynb [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+   ],
+   "source": [
+    "from sklearn.metrics import accuracy_score, precision_recall_fscore_support\n",
+    "y_true = t5_test['label']\n",
+    "y_pred = t5_test['label_nrc']\n",
+    "\n",
+    "accuracy = accuracy_score(y_true, y_pred)\n",
+    "precision, recall, f1, _ = precision_recall_fscore_support(y_true, y_pred, average='weighted')\n",
+    "\n",
+    "# show the results\n",
```
Comment (SaraHoxha) [inline NRC/t5_nrc_intensity.ipynb]: add code to save the results to a file (we shouldn't do it manually)

File: NRC/t5_nrc_intensity.ipynb [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "id": "a3f458c0-ef02-4e30-a2d0-e7562b4fc5a3",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "t5_test.to_csv('t5_nrc_inten_test.csv')"
```
Comment (SaraHoxha) [inline NRC/t5_nrc_intensity.ipynb]: again here discrepancy between xls and  csv. lets use csv

**Review by SaraHoxha — CHANGES_REQUESTED** [review]
@DucPhamBP Please look at the comments i made through the code. Most importantly, you need to change the NRC you are using for 2 reasons: 
1. You are currently using a NRC that maps to 8 emotions whereas the dataset is using Parrott's 6 emotions.  This creates a misalignment in the matching, we need to have 6 in both.
2. We are using the "Basic Emotions" approach not the multiimensional approach of emotions, which means you cannot use NRC based on Intensity but rather the original NRC which you can find at https://saifmohammad.com/WebPages/NRC-Emotion-Lexicon.htm

**Comment by DucPhamBP** [comment]
1. The NRC's emotion has 8 emotions (anger, anticipation, disgust, fear, joy, sadness, surprise, trust)  and 2 sentiments (positive, negative). The Parrott's emotion has 6 emotions (love, joy, surprise, fear, anger, sadness). There are different between 2 models. When we use NRC, we have to accept 8 emotions because the model only has them. If we reduce to 6 emotions, it is not the basic NRC model
2. I reuploaded the prediction based on the basis NRC model by your request.
3. In order to map 6 emotions as your request. I did the new file Map_NRC_to_Parrott_emotion.ipynb which I map the predicted_nrc_label to Parrott's emotions using the classification table from the Parrott's emotions by groups (https://en.wikipedia.org/wiki/Emotion_classification). I have also made some adjustments based on the meanings of certain words. Need your review

**Comment by SaraHoxha** [comment]
@DucPhamBP file paths are still not made correct, there's no need for you to have a duplicate file of the test files for yangswei and t5.

@RafaelUrbina  could you please check the mapping for Parrott and this PR in general? Thanks

**Comment by RafaelUrbina** [comment]
@SaraHoxha @DucPhamBP Some points I would change:

1.Appears that you dont need this one basic_NRC_prediction.ipynb for the mapped predictions so I wouldn't but it, if you wanna still have it add it to .gitignore

2. The test .csv files are already in the data folder, why did you bring them to the nrc folder? that makes for confusion, I would just add a relative path to the test files in the data folder and delete the test files that you manually added to the NRC folder.

3. The neutral emotion still exists and I think this will bias the predictions a lot, from what I can see in your initial prediction, I would try to find a way to do something with the neutral label honestly.

Also I think we should ask @luispinto99  and @naomiicm for their thoughts and suggestions (review) here or after you fix these points.

**Comment by DucPhamBP** [comment]
@RafaelUrbina I added the results that using NRC intensity emotions, it helps us avoid "neutral" as I told you

**Comment by SaraHoxha** [comment]
@DucPhamBP The changes you are making are not solving the issues mentioned by Rafa and I.

1. We are using the "Basic Emotions" approach not the multiimensional approach of emotions, which means you cannot use NRC based on Intensity but rather the original NRC which you can find at https://saifmohammad.com/WebPages/NRC-Emotion-Lexicon.htm . Why do you keep using intensity??
2.  basic_NRC_prediction.ipynb is not needed, yet i see it on the PR. why?
3. test sets are in the PR again but not needed, why?
4. file paths are not updated to be correct, why?
5. i see "neutral" still exists as a label, can you explain what you changed and what lines of the code you did? 

I kindly ask you to provide more descriptive comments in the commits and the comments you leave on the PR as it's not clear and it seems you are misunderstanding some of the reviews.

I remind all of us that the deadline is in 10 days.
Thank you

**Comment by DucPhamBP** [comment]
@SaraHoxha 
1. @RafaelUrbina asked to avoid "neutral" label, so I uploaded again the intensity NRC files to let everyone can review. It is just the files to everyone can review the way to avoid "neutral" label. If everyone don't accept, I will remove and make it become clean. I don't remove the "Basic Emotions", I only add the other approach to let everyone review.
2. Let everyone review then I will clean the folder in one time
3. I removed the test sets before, can you check again? I only uploaded the results dataset after mapping
5. In the intensity results, it does not has the neutral label, can you check again?

**Comment by SaraHoxha** [comment]
@DucPhamBP  Please write a comment here and tag us all when you have made the necessary changes as per the comments. the branch should be *cleaned* which means ready for merge at any moment. By clean I mean:
- only necessary notebooks
- notebooks are commented
- unnecessary code or files are deleted
- paths should be relative (not absolute) and correct

Moreover, please provide a comment where you explain the changes you made, references and how the issues are solved. Again I remind you to use the basic NRC as it aligns with our approach.

Updated deadline for this is _**Fri 3 Jan**_

Thank you

**Comment by DucPhamBP** [comment]
@SaraHoxha 
I have implemented the code using a different approach but still based on the basic NRC Lexicon as you requested. I used SBERT to convert texts into embedding vectors and also transformed the emotions in the NRC model into embeddings. Then, I calculated cosine similarity to measure and selected the label with the highest similarity. After predicting with NRC, I mapped the results to Parrott's emotions to reduce the number of emotions from 8 to 6.

This file solved your concerns as follows:
1. It uses the basic NRC Lexicon from https://saifmohammad.com/WebPages/NRC-Emotion-Lexicon.htm. 
2. It eliminates the neutral label issue. This file does not include a neutral or unpredicted label. 
3. It resolves the file path issue. I used the GitHub link directly, so you don't need to import it from a local file.

Please review and let me know your thoughts. I’ve worked hard to make this as effective as possible.

**Comment by SaraHoxha** [comment]
@DucPhamBP Thank you for updating the PR. I will review it and let you know.

@RafaelUrbina @naomiicm @luispinto99 Please review within tomorrow. Thank you

**Review by SaraHoxha — COMMENTED** [review]
@DucPhamBP Firstly, I would like to thank you for the work done as it's quite an improvement from your previous work.  I think the directory structure looks quite good and the use of the original NRC Lexicon is great.

However, I have my doubts about using SBERT for embedding the words in the lexicon. The goal of this part of the project is to compare  traditional lexicon-approaches with the models we have developed (SVM + deep learning NN). By adding this SBERT twist, we are not anymore comparing lexicon and deep learning, but more so one type of deep learning vs. another. Moreover, the preprocessing you are doing again doesn't align with the traditional lexicon approach. To me it seems you are making a mix of transformers and NRC lexicon.

Furthermore, I tried the original lexicon approach myself and I got higher results  (around 30-33%) than the one's you are getting with SBERT, which leads me to believe this "modern" approach does not justify itself.

In conclusion, I believe we should:
1. Stick to the traditional-lexicon approach to have an accurate comparison. This would mean not using SBERT embeddings, not doing preprocessing that require removal of stopwords or lemmatization.
3. Please put relative paths (not full github path, we are sending to the professor a zip file).

**Comment by RafaelUrbina** [comment]
@DucPhamBP  After reviewing the PR I have the same concerns as Sara.

1. Removal of SBERT and implementation of plain NRC

2. Change the paths to be relative (search what a relative part is)

The rest Is okay.

**Review by naomiicm — COMMENTED** [review]
Hi, these are the points @luispinto99 and I would improve:

1. Since, as the guys said, we aim compare traditional lexicon-based methods, you should stick to the original lexicon approach. It is possible to open another pipeline if you wish for the SBERT but that should be completely separated from the original one, but we don't think this is really necessary.

2. If you are going to create another pipeline with SBERT -to compare results for example- you can use the preprocessing you already have in the code, but for the original one it's not a good thing to do, in fact the results may be affected in a negative way.

3. It would be good to manually validate the labels, you could do the same thing we did before taking random samples of the predicted emotions to see if it makes sense to you. 

4. We're not entirely sure if cosine similarity is the right measure to apply if we stick to the traditional lexicon approach because it may underestimate the differences between vectors, correct us if we're wrong.

That would be all from me and @luispinto99 , in addition to the comments already made by @SaraHoxha and @RafaelUrbina .

**Comment by SaraHoxha** [comment]
@DucPhamBP Based on review comments by teammates, please push the changes by tomorrow noon. Thank you

**Comment by DucPhamBP** [comment]
@SaraHoxha 
I solved as you want:
1. I used the tradition NRC Lexicon approach.
2. I eliminated the "neutral" label.
3. The accuracy is improved.
4. I used relative paths.
5. I didn't use SBERT.
Answer to @naomiicm, cosine similarity is commonly used to measure the similarity between two contextual text embeddings. It calculates the cosine of the angle between two vectors (embeddings) in a high-dimensional space, which helps determine how similar they are regardless of their magnitude.. I referred from some papers https://arxiv.org/pdf/2410.01306, https://aclanthology.org/2021.wassa-1.22.pdf?utm_source=chatgpt.com, https://www.sciencedirect.com/science/article/pii/S2949719124000530?ref=pdf_download&fr=RR-2&rr=8f67d8510c41eea9#b37, I tried to take a test with this approach, they both do with NRC and other lexicons.

**Comment by SaraHoxha** [comment]
@DucPhamBP the paths were still not correct, so I fixed them. the rest looks ok. I'm approving the PR. Remember you have to write the report portion for NRC (methods, results, discussion).


@naomiicm please review and approve/reject accordingly so we can proceed with merge.

**Review by SaraHoxha — APPROVED** [review]

**Review by naomiicm — APPROVED** [review]
